src/base/bittorrent/session.cpp

```diff
@@ -3831,13 +3831,18 @@ void Session::handleTorrentMetadataReceived(TorrentHandle *const torrent)
     torrent->saveResumeData();
 
     // Save metadata
-    const QDir resumeDataDir(m_resumeFolderPath);
-    QString torrentFile = resumeDataDir.absoluteFilePath(QString("%1.torrent").arg(torrent->hash()));
-    if (torrent->saveTorrentFile(torrentFile)) {
+    const QDir resumeDataDir {m_resumeFolderPath};
+    const QString torrentFileName {QString {"%1.torrent"}.arg(torrent->hash())};
+    try {
+        torrent->saveTorrentFile(resumeDataDir.absoluteFilePath(torrentFileName));
         // Copy the torrent file to the export folder
         if (!torrentExportDirectory().isEmpty())
             exportTorrentFile(torrent);
     }
+    catch (const RuntimeError &err) {
```

Comment:
General question, we don't throw/catch exceptions in these classes, should we start doing that now?
IMO it is a bit troublesome that `TorrentHandle::saveTorrentFile()` could propagate exceptions, I mean these is no easy way to know that when writing code calling that function without a deeper dive into its implementation.

Reply:
>General question, we don't throw/catch exceptions in these classes, should we start doing that now?

I am now convinced that exceptions are the best way to handle errors in such cases (especially when it comes to passing an error through multiple levels).

The only problem (you're right) is to specify that the function can throw an exception. As far as I understand, there is no special way here other than just to document this fact.

Otherwise we need additional parameter to manage error information.

Reply:
What about repeating the try/catch code in `TorrentHandle::saveTorrentFile()`?
```c++
void TorrentHandle::saveTorrentFile(const QString &path) const
{
    try {
        m_torrentInfo.saveToFile(path);
    }
    catch (const RuntimeError &err) {
        throw err;
    }
}
```
It is dumb but at least it is clear that it will throw exception.

Reply:
It still requires seeing the source code. Also it add some additional code in each level so it negates all the benefits...

Reply:
>there is no special way here other than just to document this fact.

According to specifications any function in c++ can potentially throw an exception unless you mark it as "noexcept". But you can mark function as potentially throwing explicitly with "noexcept(false)".

Reply:
>But you can mark function as potentially throwing explicitly with "noexcept(false)".

Interesting, I think we can add it to TorrentHandle::saveTorrentFile().

Reply:
Another question: do we really need `TorrentHandle::saveTorrentFile()` wrapper?

Reply:
>Another question: do we really need TorrentHandle::saveTorrentFile() wrapper?

If `saveTorrentFile()` really belongs to TorrentInfo class then there is no point of having a wrapper.

Reply:
>>But you can mark function as potentially throwing explicitly with "noexcept(false)".
>
>Interesting, I think we can add it to TorrentHandle::saveTorrentFile().


My subsequent thinking led me to the conclusion that we should not do it this way.
`noexcept` with parameter has a different purpose (instantiating different versions of functions from the template etc).
Although it is not forbidden to use it explicitly, it does not do any good. On the contrary, it turns out that it can mislead all other functions (not having this statement) that they are exception-safe" (although this is not the case, even if we don't explicitly throw exceptions inside them).